src/tree_sortable.rs

```diff
@@ -0,0 +1,106 @@
+// This file was generated by gir (17af302) from gir-files (11e0e6d)
+// DO NOT EDIT
+
+use SortType;
+use ffi;
+use glib::object::IsA;
+use glib::translate::*;
+use auto::traits::TreeSortableExt as Auto;
+use std::thread;
+use std::cell::RefCell;
+use std::mem::transmute;
+use std::process;
+
+use glib_ffi::gpointer;
+use {TreeSortable, TreeIter};
+use ffi::{GtkTreeIter, GtkTreeSortable};
+
+pub trait TreeSortableExt {
+    // -- manual --
+    fn set_default_sort_func<F>(&self, sort_func: F)
+    where F: FnMut(&TreeSortable, &TreeIter, &TreeIter) -> i32 + 'static;
+    fn set_sort_func<F>(&self, sort_column_id: i32, sort_func: F)
+    where F: FnMut(&TreeSortable, &TreeIter, &TreeIter) -> i32 + 'static;
+    // -- auto --
+    fn get_sort_column_id(&self) -> Option<(i32, SortType)>;
+    fn has_default_sort_func(&self) -> bool;
+    fn set_sort_column_id(&self, sort_column_id: i32, order: SortType);
+    fn sort_column_changed(&self);
+}
+
+struct CallbackGuard;
+
+impl Drop for CallbackGuard {
+    fn drop(&mut self) {
+        if thread::panicking() {
+            process::exit(101);
+        }
+    }
+}
+
+macro_rules! callback_guard {
+    () => (
+        let _guard = CallbackGuard;
+        if cfg!(debug_assertions) {
+            assert_initialized_main_thread!();
+        }
+    )
+}
+
+unsafe extern "C" fn trampoline(this: *mut GtkTreeSortable, iter: *mut GtkTreeIter,
+        iter2: *mut GtkTreeIter, f: &mut Box<FnMut(&TreeSortable, &TreeIter, &TreeIter) -> i32 + 'static>) {
+    callback_guard!();
+    f(&from_glib_none(this), &from_glib_borrow(iter), &from_glib_borrow(iter2));
+}
+
+unsafe extern "C" fn destroy_closure(ptr: gpointer) {
+    callback_guard!();
+    Box::<RefCell<Box<FnMut(&TreeSortable, &TreeIter, &TreeIter) -> i32 + 'static>>>::from_raw(ptr as *mut _);
+}
+
+fn into_raw<F: FnMut(&TreeSortable, &TreeIter, &TreeIter) -> i32 + 'static>(func: F) -> gpointer {
+    let func: Box<RefCell<Box<FnMut(&TreeSortable, &TreeIter, &TreeIter) -> i32 + 'static>>> =
+        Box::new(RefCell::new(Box::new(func)));
+    Box::into_raw(func) as gpointer
+}
+
+impl<O: IsA<TreeSortable>> TreeSortableExt for O {
+    #[inline]
+    fn get_sort_column_id(&self) -> Option<(i32, SortType)> {
+        Auto::get_sort_column_id(self)
+    }
+
+    #[inline]
+    fn has_default_sort_func(&self) -> bool {
+        Auto::has_default_sort_func(self)
+    }
+
+    fn set_default_sort_func<F>(&self, sort_func: F)
+        where F: FnMut(&TreeSortable, &TreeIter, &TreeIter) -> i32 + 'static {
+        unsafe {
+            ffi::gtk_tree_sortable_set_default_sort_func(self.to_glib_none().0,
+                                                         transmute(trampoline),
```

Comment:
I used to do transmutes like this myself but now I'd prefer to move to a different convention: make the trampoline's signature compatible with this function (`user_data: gpointer`) and use either `transmute` or `as` in the trampoline itself.